### PR TITLE
Block2 DMRG Solver

### PR DESCRIPTION
**Pull request by hczhai**
## Description
1. Added the ``block2`` DMRG solver.
2. Fixed a bug in ``RDMsSpinDependent::SF_G2()``.
3. Added ``from __future__ import annotations`` in ``forte/data.py`` to make it work with Python 3.7.

## User Notes
- [x] Features added

- Added a DMRG solver interfaced with ``block2``. Minimal required ``block2`` version is 0.5.3rc7. Installation and usage documentation can be found in: https://block2.readthedocs.io/en/latest/user/forte.html.

- [x] Changes to compilation

- Added a compilation option ``ENABLE_block2``. Default is ``OFF``.

## Checklist
- [x] Removed comments in code and input files
- [x] Documented source code
- [x] Checked for redundant headers
- [x] Ready to go!

### CODE REVIEW

**Review by lcyyork — COMMENTED** [review]
This is a great feature to have. May I ask for a test case and reference output?

**Comment by fevangelista** [comment]
Thank you for adding this new feature! I'd like to give it some testing before committing it. At some point, the interface will have to change to use the new module system, but that can wait until the interface design is settled.

**Comment by lcyyork** [comment]
In two-step MCSCF, there is a mismatch between the DMRG energy and that reconstructed using DMRG RDMs unless they are in the same canonicalization. Line 260 of `mcscf_2step.cc` may be changed to:
```
bool skip_de_conv = (ci_type_.find("DMRG") != std::string::npos or
                             ci_type_.find("BLOCK2") != std::string::npos);
```

**Comment by hczhai** [comment]
> Line 260 of mcscf_2step.cc may be changed to:

Thanks for pointing this out. This is now added.

**Comment by fevangelista** [comment]
I'll merge this after the last PR changes pass. Let me know if there are other reasons to hold this off.

**Review by lcyyork — APPROVED** [review]

File: forte/dmrg/block2_dmrg_solver.cc
Comment (lcyyork) [inline forte/dmrg/block2_dmrg_solver.cc:370]: I wonder is there a `dump_wfn` implementation available? It seems to me that every DMRG part of the two-step MCSCF procedure always starts from a random initial guess, which leads to oscillating energies even when the orbitals are well converged.

Comment (hczhai) [reply forte/dmrg/block2_dmrg_solver.cc:370]: The MPS is automatically stored in scratch storage so there is no need to explicitly dump the MPS. One only needs to set this ``READ_ACTIVE_WFN_GUESS`` option (when it is close to convergence) to reuse the MPS from a previous MCSCF step.

**Comment by fevangelista** [comment]
It looks like this PR needs to be changed to address the new modifications in Forte. @hczhai could you take a look? I would be happy to help in case some of the changes are confusing.

**Comment by fevangelista** [comment]
Thank you so much for this contribution @hczhai!

**Comment by hczhai** [comment]
Thanks for merging this! @lcyyork helped a lot in testing and improving this interface.

Actually I have a small additional change https://github.com/hczhai/forte/commit/df179ecb5fd90839f9dcfcb9358c895b15f82fe2 to prevent some future accidental temporary change in block2 source code from breaking the testing pipelines here.

**Pull request merged**
